File: code/gui/hmi/programs/calib_robot.mjs
import { shallowRef, ref } from 'vue'
import { url, poll } from 'web/utils'
import { ButtonBar, feedback } from 'web/widgets'
import { hmiIndex } from 'hmi'



hmiIndex.addPage('calib_robot', {
	targetGuard: 'app@calib_robot',
	setup() {

		const drivesUrl = url('calib_robot', 'app');

		const calibPos = ref(310);
		const drives = shallowRef([]);

		const drivesPoll = poll(500, async ()=>{
			drives.value = await drivesUrl.fetchJson();
		});

		async function save(ev, name) {
			await feedback(ev.target, drivesUrl.postJson({ name, calibPos:calibPos.value }));
			drivesPoll();
		}

		return { calibPos, drives, save };
	},
	template: //html
	`
	<div class="mb-4">
		<label class="form-label">{{ $t('hmi.calib_robot.calibPos') }}</label>
		<input v-model.number="calibPos" type="number" step="any" class="form-control">
	</div>
	<div v-for="drive in drives" :key="drive.name" class="mb-3">
		<label class="form-label">{{ drive.name }}</label>
		<div class="row gy-2">
			<div class="col-lg input-group">
				<span class="input-group-text">{{ $t('hmi.calib_robot.encoder') }}</span>
				<input :value="drive.encoder.toFixed(2)" readonly class="form-control">
			</div>
			<div class="col-lg input-group">
				<span class="input-group-text">{{ $t('hmi.calib_robot.offset') }}</span>
				<input :value="(calibPos - drive.encoder).toFixed(2)" readonly class="form-control">
			</div>
			<div class="col-lg input-group">
				<span class="input-group-text">{{ $t('hmi.calib_robot.savedOffset') }}</span>
				<input :value="drive.offset.toFixed(2)" readonly class="form-control">
			</div>
			<div class="col-lg-auto">
				<button @click="save($event, drive.name)" class="btn btn-primary">{{ $t('save') }}</button>
			</div>
		</div>
	</div>
	`
})

File: code/gui/hmi/programs/calib_robot.py
import web
from shared.conf import Conf
from shared.iceoryx.pubsub import IoxSubscriber, poll
from shared.topics.programs import calib_robot_pubsub


web.document.imports.append('hmi/programs/calib_robot')
web.site.show(__name__, lambda: calib_robot_pubsub.dynamic_config.number_of_publishers)



@web.handler
class drives(web.RequestHandler):

	conf = Conf('/etc/app/drives.conf')
	calib_subscriber = IoxSubscriber(calib_robot_pubsub)

	async def prepare(self):
		self.encoders:dict = await poll(self.calib_subscriber.receive_msgpack)

	async def get(self):
		self.write([{
			'name': name,
			'encoder': encoder,
			'offset': self.conf.getfloat(name, 'offset', fallback=0.0),
		} for name, encoder in self.encoders.items()])

	async def post(self):
		drive = self.read_json()
		offset = drive['calibPos'] - self.encoders[drive['name']]
		self.conf.update({ drive['name']: { 'offset': offset } })
		self.conf.save()
